include/realizations/catchment/Bmi_Module_Formulation.hpp

```diff
@@ -306,9 +306,9 @@ namespace realization {
          *
          * @return Shared pointer to the backing model object that implements the BMI.
          */
-        std::shared_ptr<models::bmi::Bmi_Adapter> get_bmi_model();
+        std::shared_ptr<models::bmi::Bmi_Adapter> get_bmi_model() const;
```

Comment:
This change I'm not entirely sure on -- it's needed for some of the other const qualified members but, I think to be correctly const qualified, the return type should be

```cpp
std::shared_ptr<const models::bmi::Bmi_Adapter>
```

But is that a possible change? Need to check

Reply:
`std::shared_ptr<const Bmi_Adapter>` is valid (which should've been obvious) -- however, it's not possible to use this result in other const qualified members because the C++ BMI interface does not const qualify its virtual functions, i.e. `bmi->GetOutputVarNames()` is not `const`.

In order for this to work, it'd require changes in the BMI header. An alternative is for member functions to use the `bmi_model` member directly, which circumvents this, but requires more changes... or just leave it and accept it as is 🥲

Reply:
I think it's fine to leave `get_bmi_model()` as non-`const` - it's `protected`, so it's not like we need to reason about unrelated classes messing with things.

The idiomatic C++ way to address this is an overload pair along the lines of 
```c++
std::shared_ptr<models::bmi::Bmi_Adapter> get_bmi_model();
std::shared_ptr<const models::bmi::Bmi_Adapter> get_bmi_model() const;
```
Obviously, that doesn't actually work in this case because that `const Bmi_Adapter` is pretty useless.

Reply:
Was `get_available_variable_names` the only place that the limitation on use of `get_bmi_model` came up? If so, great.

Reply:
No, it's used in quite a few functions intended to be `const`. I think it may easy enough to leave `get_bmi_model()` as non-`const` and replace them with direct member access, since most of the calls are of the form `get_bmi_model()->GetSomething()`.

Reply:
It looks like the codebase was ultimately happy with this revision as made. I don't see any collateral damage changes in the PR, and it compiled (and passed the tests).

Reply:
If you're worried about "`get_bmi_model()` should be deeply-`const` but that's infeasible", I wouldn't fret over it.

Reply:
Yeah, that was my only concern. It irks me a bit, but I guess there's no real harm in it 😅